src/main/java/com/icthh/xm/uaa/security/oauth2/idp/IdpTokenGranter.java

```diff
@@ -0,0 +1,226 @@
+import com.icthh.xm.uaa.domain.UserLoginType;
+import com.icthh.xm.uaa.domain.properties.TenantProperties;
+import com.icthh.xm.uaa.security.DomainUserDetails;
+import com.icthh.xm.uaa.security.DomainUserDetailsService;
+import com.icthh.xm.uaa.security.oauth2.idp.source.model.XmAuthenticationToken;
+import com.icthh.xm.uaa.service.TenantPropertiesService;
+import com.icthh.xm.uaa.service.UserLoginService;
+import com.icthh.xm.uaa.service.UserService;
+import com.icthh.xm.uaa.service.dto.UserDTO;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.security.authentication.AuthenticationServiceException;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.mapping.GrantedAuthoritiesMapper;
+import org.springframework.security.core.authority.mapping.NullAuthoritiesMapper;
+import org.springframework.security.oauth2.common.OAuth2AccessToken;
+import org.springframework.security.oauth2.provider.ClientDetails;
+import org.springframework.security.oauth2.provider.ClientDetailsService;
+import org.springframework.security.oauth2.provider.OAuth2Authentication;
+import org.springframework.security.oauth2.provider.OAuth2RequestFactory;
+import org.springframework.security.oauth2.provider.TokenRequest;
+import org.springframework.security.oauth2.provider.token.AbstractTokenGranter;
+import org.springframework.security.oauth2.provider.token.AuthorizationServerTokenServices;
+
+import java.util.Collection;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * This class responsible for handling requests with idp token
+ */
+@Slf4j
+@LepService(group = "security.idp")
+public class IdpTokenGranter extends AbstractTokenGranter {
+
+    public static final String GRANT_TYPE_IDP_TOKEN = "idp_token";
+
+    private static final String FIRST_NAME_CONFIG_ATTRIBUTE = "firstNameAttribute";
+    private static final String LAST_NAME_CONFIG_ATTRIBUTE = "lastNameAttribute";
+    private static final String USER_IDENTITY_CONFIG_TYPE = "userIdentityType";
+    private static final String USER_IDENTITY_CONFIG_ATTRIBUTE = "userIdentityAttribute";
+
+    private final XmJwkTokenStore jwkTokenStore;
+    private final DomainUserDetailsService domainUserDetailsService;
+    private final TenantPropertiesService tenantPropertiesService;
+    private final UserService userService;
+    private final UserLoginService userLoginService;
+    private final GrantedAuthoritiesMapper authoritiesMapper = new NullAuthoritiesMapper();
+
+    public IdpTokenGranter(AuthorizationServerTokenServices tokenServices,
+                           ClientDetailsService clientDetailsService,
+                           @Lazy OAuth2RequestFactory requestFactory,
```

Comment:
check if the instance will be created in config method with @Bean then this @Lazy probably will not need